File: frontend/src/utils/permissions.ts
export const VALID_ROLES = [
  'ADMIN',
  'PROJECT_OWNER',
  'BUSINESS_ANALYST',
  'SOLUTION_ARCHITECT',
  'MANAGER',
  'MEMBER',
  'VIEWER'
] as const;

export type SystemRole = typeof VALID_ROLES[number];

// Centralized Role Permissions Map
export const ROLE_PERMISSIONS: Record<string, string[]> = {
  // 1. ADMIN
  ADMIN: [
    'admin.access',
    'organization.manage',
    'workspace.manage',
    'user.manage',
    'audit.view',
    'project.view',
    'project.manage',
    'team.manage',
    'document.view',
    'document.upload',
    'document.delete',
    'version.view',
    'export.create'
  ],

  // 2. PROJECT_OWNER
  PROJECT_OWNER: [
    'workspace.manage',
    'project.view',
    'project.create',
    'project.edit',
    'project.delete',
    'project.manage',
    'team.manage',
    'document.view',
    'document.upload',
    'document.delete',
    'discovery.chat',
    'requirements.view',
    'requirements.edit',
    'stakeholders.view',
    'stakeholders.edit',
    'analysis.view',
    'analysis.run',
    'analysis.edit',
    'gap_analysis.view',
    'gap_analysis.run',
    'gap_analysis.edit',
    'recommendation.view',
    'recommendation.generate',
    'recommendation.edit',
    'recommendation.approve',
    'process.view',
    'process.generate',
    'process.edit',
    'architecture.view',
    'architecture.generate',
    'architecture.edit',
    'database.view',
    'database.generate',
    'database.edit',
    'api.view',
    'api.generate',
    'api.edit',
    'ux.view',
    'ux.generate',
    'ux.edit',
    'planning.view',
    'planning.generate',
    'planning.edit',
    'risk.view',
    'risk.edit',
    'score.view',
    'score.calculate',
    'simulation.run',
    'blueprint.view',
    'blueprint.generate',
    'blueprint.edit',
    'blueprint.approve',
    'export.create',
    'comment.create',
    'version.view',
    'version.create',
    'audit.view'
  ],

  // 3. BUSINESS_ANALYST
  BUSINESS_ANALYST: [
    'project.view',
    'document.view',
    'document.upload',
    'discovery.chat',
    'requirements.view',
    'requirements.edit',
    'stakeholders.view',
    'stakeholders.edit',
    'analysis.view',
    'analysis.run',
    'analysis.edit',
    'gap_analysis.view',
    'gap_analysis.run',
    'gap_analysis.edit',
    'recommendation.view',
    'recommendation.generate',
    'recommendation.review_ba',
    'process.view',
    'process.generate',
    'process.edit',
    'architecture.view',
    'database.view',
    'api.view',
    'ux.view',
    'ux.generate',
    'ux.edit',
    'planning.view',
    'planning.generate',
    'risk.view',
    'risk.edit',
    'score.view',
    'score.calculate',
    'simulation.run',
    'blueprint.view',
    'export.create',
    'comment.create',
    'version.view'
  ],

  // 4. SOLUTION_ARCHITECT
  SOLUTION_ARCHITECT: [
    'project.view',
    'document.view',
    'document.upload',
    'discovery.chat',
    'requirements.view',
    'stakeholders.view',
    'analysis.view',
    'gap_analysis.view',
    'recommendation.view',
    'recommendation.generate',
    'recommendation.review_arch',
    'process.view',
    'process.generate',
    'process.edit',
    'architecture.view',
    'architecture.generate',
    'architecture.edit',
    'architecture.submit',
    'database.view',
    'database.generate',
    'database.edit',
    'api.view',
    'api.generate',
    'api.edit',
    'ux.view',
    'ux.generate',
    'planning.view',
    'planning.generate',
    'risk.view',
    'risk.edit',
    'score.view',
    'score.calculate',
    'simulation.run',
    'blueprint.view',
    'export.create',
    'comment.create',
    'version.view',
    'version.create'
  ],

  // 5. MANAGER
  MANAGER: [
    'project.view',
    'document.view',
    'document.upload',
    'discovery.chat',
    'requirements.view',
    'stakeholders.view',
    'analysis.view',
    'gap_analysis.view',
    'recommendation.view',
    'recommendation.approve',
    'process.view',
    'architecture.view',
    'database.view',
    'api.view',
    'ux.view',
    'planning.view',
    'risk.view',
    'score.view',
    'simulation.run',
    'blueprint.view',
    'blueprint.approve',
    'export.create',
    'comment.create',
    'version.view',
    'audit.view'
  ],

  // 6. MEMBER
  MEMBER: [
    'project.view',
    'document.view',
    'document.upload',
    'discovery.chat',
    'requirements.view',
    'stakeholders.view',
    'analysis.view',
    'gap_analysis.view',
    'recommendation.view',
    'process.view',
    'architecture.view',
    'database.view',
    'api.view',
    'ux.view',
    'planning.view',
    'risk.view',
    'score.view',
    'simulation.run',
    'blueprint.view',
    'export.create',
    'comment.create',
    'version.view'
  ],

  // 7. VIEWER (Read-Only)
  VIEWER: [
    'project.view',
    'document.view',
    'discovery.chat',
    'requirements.view',
    'stakeholders.view',
    'analysis.view',
    'gap_analysis.view',
    'recommendation.view',
    'process.view',
    'architecture.view',
    'database.view',
    'api.view',
    'ux.view',
    'planning.view',
    'risk.view',
    'score.view',
    'blueprint.view',
    'export.create',
    'version.view'
  ]
};

// Aliases
ROLE_PERMISSIONS.OWNER = ROLE_PERMISSIONS.PROJECT_OWNER;
ROLE_PERMISSIONS.ANALYST = ROLE_PERMISSIONS.BUSINESS_ANALYST;
ROLE_PERMISSIONS.ARCHITECT = ROLE_PERMISSIONS.SOLUTION_ARCHITECT;

/**
 * Check if a role has a specific permission
 */
export function hasPermission(role: string, permissionKey: string): boolean {
  if (!role) return false;
  const roleKey = role.toUpperCase();
  const perms = ROLE_PERMISSIONS[roleKey] || [];
  return perms.includes(permissionKey);
}

/**
 * Check if a role has ANY of the specified permissions
 */
export function hasAnyPermission(role: string, permissionKeys: string[]): boolean {
  if (!role || !permissionKeys.length) return false;
  const roleKey = role.toUpperCase();
  const perms = ROLE_PERMISSIONS[roleKey] || [];
  return permissionKeys.some(key => perms.includes(key));
}

/**
 * Check if a role has ALL of the specified permissions
 */
export function hasAllPermissions(role: string, permissionKeys: string[]): boolean {
  if (!role || !permissionKeys.length) return false;
  const roleKey = role.toUpperCase();
  const perms = ROLE_PERMISSIONS[roleKey] || [];
  return permissionKeys.every(key => perms.includes(key));
}

/**
 * Helper to check view permission for a module
 */
export function canView(role: string, moduleName: string): boolean {
  return hasPermission(role, `${moduleName}.view`);
}

/**
 * Helper to check edit permission for a module
 */
export function canEdit(role: string, moduleName: string): boolean {
  return hasPermission(role, `${moduleName}.edit`);
}

/**
 * Helper to check create permission for a module
 */
export function canCreate(role: string, moduleName: string): boolean {
  return hasPermission(role, `${moduleName}.create`) || hasPermission(role, `${moduleName}.generate`);
}

/**
 * Helper to check delete permission for a module
 */
export function canDelete(role: string, moduleName: string): boolean {
  return hasPermission(role, `${moduleName}.delete`);
}

/**
 * Helper to check approve permission for a module
 */
export function canApprove(role: string, moduleName: string): boolean {
  return hasPermission(role, `${moduleName}.approve`);
}

/**
 * Check if role is read-only (e.g. VIEWER)
 */
export function isReadOnlyRole(role: string): boolean {
  return role?.toUpperCase() === 'VIEWER';
}

File: backend/app/auth/permissions.py
from typing import Dict, Set, Optional, List, Union
from enum import Enum

class Permission(str, Enum):
    # Admin / Platform
    ADMIN_ACCESS = "admin.access"
    ORGANIZATION_MANAGE = "organization.manage"
    WORKSPACE_MANAGE = "workspace.manage"
    USER_MANAGE = "user.manage"
    AUDIT_VIEW = "audit.view"
    
    # Project Core
    PROJECT_VIEW = "project.view"
    PROJECT_CREATE = "project.create"
    PROJECT_EDIT = "project.edit"
    PROJECT_DELETE = "project.delete"
    PROJECT_MANAGE = "project.manage"
    TEAM_MANAGE = "team.manage"
    
    # Documents
    DOCUMENT_VIEW = "document.view"
    DOCUMENT_UPLOAD = "document.upload"
    DOCUMENT_DELETE = "document.delete"
    
    # Business Analysis & Discovery
    DISCOVERY_CHAT = "discovery.chat"
    REQUIREMENTS_VIEW = "requirements.view"
    REQUIREMENTS_EDIT = "requirements.edit"
    STAKEHOLDERS_VIEW = "stakeholders.view"
    STAKEHOLDERS_EDIT = "stakeholders.edit"
    ANALYSIS_VIEW = "analysis.view"
    ANALYSIS_RUN = "analysis.run"
    ANALYSIS_EDIT = "analysis.edit"
    GAP_ANALYSIS_VIEW = "gap_analysis.view"
    GAP_ANALYSIS_RUN = "gap_analysis.run"
    GAP_ANALYSIS_EDIT = "gap_analysis.edit"
    
    # Solutions & Recommendations
    RECOMMENDATION_VIEW = "recommendation.view"
    RECOMMENDATION_GENERATE = "recommendation.generate"
    RECOMMENDATION_EDIT = "recommendation.edit"
    RECOMMENDATION_REVIEW_BA = "recommendation.review_ba"
    RECOMMENDATION_REVIEW_ARCH = "recommendation.review_arch"
    RECOMMENDATION_APPROVE = "recommendation.approve"
    
    # Technical Architecture & Engineering
    PROCESS_VIEW = "process.view"
    PROCESS_GENERATE = "process.generate"
    PROCESS_EDIT = "process.edit"
    
    ARCHITECTURE_VIEW = "architecture.view"
    ARCHITECTURE_GENERATE = "architecture.generate"
    ARCHITECTURE_EDIT = "architecture.edit"
    ARCHITECTURE_SUBMIT = "architecture.submit"
    
    DATABASE_VIEW = "database.view"
    DATABASE_GENERATE = "database.generate"
    DATABASE_EDIT = "database.edit"
    
    API_VIEW = "api.view"
    API_GENERATE = "api.generate"
    API_EDIT = "api.edit"
    
    UX_VIEW = "ux.view"
    UX_GENERATE = "ux.generate"
    UX_EDIT = "ux.edit"
    
    # Planning & Scoring
    PLANNING_VIEW = "planning.view"
    PLANNING_GENERATE = "planning.generate"
    PLANNING_EDIT = "planning.edit"
    
    RISK_VIEW = "risk.view"
    RISK_EDIT = "risk.edit"
    
    SCORE_VIEW = "score.view"
    SCORE_CALCULATE = "score.calculate"
    SIMULATION_RUN = "simulation.run"
    
    # Blueprint & Governance
    BLUEPRINT_VIEW = "blueprint.view"
    BLUEPRINT_GENERATE = "blueprint.generate"
    BLUEPRINT_EDIT = "blueprint.edit"
    BLUEPRINT_APPROVE = "blueprint.approve"
    EXPORT_CREATE = "export.create"
    
    # Collaboration
    COMMENT_CREATE = "comment.create"
    VERSION_VIEW = "version.view"
    VERSION_CREATE = "version.create"

# Strict Page/Module Access Matrix as defined in Section 10 of Hackathon Specification
ROLE_PERMISSIONS: Dict[str, Set[str]] = {
    # 1. ADMIN
    "ADMIN": {
        Permission.ADMIN_ACCESS.value,
        Permission.ORGANIZATION_MANAGE.value,
        Permission.WORKSPACE_MANAGE.value,
        Permission.USER_MANAGE.value,
        Permission.AUDIT_VIEW.value,
        Permission.PROJECT_VIEW.value,
        Permission.PROJECT_MANAGE.value,
        Permission.TEAM_MANAGE.value,
        Permission.DOCUMENT_VIEW.value,
        Permission.DOCUMENT_UPLOAD.value,
        Permission.DOCUMENT_DELETE.value,
        Permission.VERSION_VIEW.value,
        Permission.EXPORT_CREATE.value,
    },
    
    # 2. PROJECT_OWNER
    "PROJECT_OWNER": {
        Permission.WORKSPACE_MANAGE.value,
        Permission.PROJECT_VIEW.value,
        Permission.PROJECT_CREATE.value,
        Permission.PROJECT_EDIT.value,
        Permission.PROJECT_DELETE.value,
        Permission.PROJECT_MANAGE.value,
        Permission.TEAM_MANAGE.value,
        Permission.DOCUMENT_VIEW.value,
        Permission.DOCUMENT_UPLOAD.value,
        Permission.DOCUMENT_DELETE.value,
        Permission.DISCOVERY_CHAT.value,
        Permission.REQUIREMENTS_VIEW.value,
        Permission.REQUIREMENTS_EDIT.value,
        Permission.STAKEHOLDERS_VIEW.value,
        Permission.STAKEHOLDERS_EDIT.value,
        Permission.ANALYSIS_VIEW.value,
        Permission.ANALYSIS_RUN.value,
        Permission.ANALYSIS_EDIT.value,
        Permission.GAP_ANALYSIS_VIEW.value,
        Permission.GAP_ANALYSIS_RUN.value,
        Permission.GAP_ANALYSIS_EDIT.value,
        Permission.RECOMMENDATION_VIEW.value,
        Permission.RECOMMENDATION_GENERATE.value,
        Permission.RECOMMENDATION_EDIT.value,
        Permission.RECOMMENDATION_APPROVE.value,
        Permission.PROCESS_VIEW.value,
        Permission.PROCESS_GENERATE.value,
        Permission.PROCESS_EDIT.value,
        Permission.ARCHITECTURE_VIEW.value,
        Permission.ARCHITECTURE_GENERATE.value,
        Permission.ARCHITECTURE_EDIT.value,
        Permission.DATABASE_VIEW.value,
        Permission.DATABASE_GENERATE.value,
        Permission.DATABASE_EDIT.value,
        Permission.API_VIEW.value,
        Permission.API_GENERATE.value,
        Permission.API_EDIT.value,
        Permission.UX_VIEW.value,
        Permission.UX_GENERATE.value,
        Permission.UX_EDIT.value,
        Permission.PLANNING_VIEW.value,
        Permission.PLANNING_GENERATE.value,
        Permission.PLANNING_EDIT.value,
        Permission.RISK_VIEW.value,
        Permission.RISK_EDIT.value,
        Permission.SCORE_VIEW.value,
        Permission.SCORE_CALCULATE.value,
        Permission.SIMULATION_RUN.value,
        Permission.BLUEPRINT_VIEW.value,
        Permission.BLUEPRINT_GENERATE.value,
        Permission.BLUEPRINT_EDIT.value,
        Permission.BLUEPRINT_APPROVE.value,
        Permission.EXPORT_CREATE.value,
        Permission.COMMENT_CREATE.value,
        Permission.VERSION_VIEW.value,
        Permission.VERSION_CREATE.value,
        Permission.AUDIT_VIEW.value,
    },
    
    # 3. BUSINESS_ANALYST (BA)
    "BUSINESS_ANALYST": {
        Permission.PROJECT_VIEW.value,
        Permission.DOCUMENT_VIEW.value,
        Permission.DOCUMENT_UPLOAD.value,
        Permission.DISCOVERY_CHAT.value,
        Permission.REQUIREMENTS_VIEW.value,
        Permission.REQUIREMENTS_EDIT.value,
        Permission.STAKEHOLDERS_VIEW.value,
        Permission.STAKEHOLDERS_EDIT.value,
        Permission.ANALYSIS_VIEW.value,
        Permission.ANALYSIS_RUN.value,
        Permission.ANALYSIS_EDIT.value,
        Permission.GAP_ANALYSIS_VIEW.value,
        Permission.GAP_ANALYSIS_RUN.value,
        Permission.GAP_ANALYSIS_EDIT.value,
        Permission.RECOMMENDATION_VIEW.value,
        Permission.RECOMMENDATION_GENERATE.value,
        Permission.RECOMMENDATION_REVIEW_BA.value,
        Permission.PROCESS_VIEW.value,
        Permission.PROCESS_GENERATE.value,
        Permission.PROCESS_EDIT.value,
        Permission.ARCHITECTURE_VIEW.value, # Read-only
        Permission.DATABASE_VIEW.value,     # Read-only
        Permission.API_VIEW.value,          # Read-only
        Permission.UX_VIEW.value,
        Permission.UX_GENERATE.value,
        Permission.UX_EDIT.value,
        Permission.PLANNING_VIEW.value,
        Permission.PLANNING_GENERATE.value,
        Permission.RISK_VIEW.value,
        Permission.RISK_EDIT.value,
        Permission.SCORE_VIEW.value,
        Permission.SCORE_CALCULATE.value,
        Permission.SIMULATION_RUN.value,
        Permission.BLUEPRINT_VIEW.value,    # Read-only
        Permission.EXPORT_CREATE.value,
        Permission.COMMENT_CREATE.value,
        Permission.VERSION_VIEW.value,
    },
    
    # 4. SOLUTION_ARCHITECT (ARCH)
    "SOLUTION_ARCHITECT": {
        Permission.PROJECT_VIEW.value,
        Permission.DOCUMENT_VIEW.value,
        Permission.DOCUMENT_UPLOAD.value,
        Permission.DISCOVERY_CHAT.value,
        Permission.REQUIREMENTS_VIEW.value,
        Permission.STAKEHOLDERS_VIEW.value,
        Permission.ANALYSIS_VIEW.value,
        Permission.GAP_ANALYSIS_VIEW.value,
        Permission.RECOMMENDATION_VIEW.value,
        Permission.RECOMMENDATION_GENERATE.value,
        Permission.RECOMMENDATION_REVIEW_ARCH.value,
        Permission.PROCESS_VIEW.value,
        Permission.PROCESS_GENERATE.value,
        Permission.PROCESS_EDIT.value,
        Permission.ARCHITECTURE_VIEW.value,
        Permission.ARCHITECTURE_GENERATE.value,
        Permission.ARCHITECTURE_EDIT.value,
        Permission.ARCHITECTURE_SUBMIT.value,
        Permission.DATABASE_VIEW.value,
        Permission.DATABASE_GENERATE.value,
        Permission.DATABASE_EDIT.value,
        Permission.API_VIEW.value,
        Permission.API_GENERATE.value,
        Permission.API_EDIT.value,
        Permission.UX_VIEW.value,
        Permission.UX_GENERATE.value,
        Permission.PLANNING_VIEW.value,
        Permission.PLANNING_GENERATE.value,
        Permission.RISK_VIEW.value,
        Permission.RISK_EDIT.value,
        Permission.SCORE_VIEW.value,
        Permission.SCORE_CALCULATE.value,
        Permission.SIMULATION_RUN.value,
        Permission.BLUEPRINT_VIEW.value,    # Read-only
        Permission.EXPORT_CREATE.value,
        Permission.COMMENT_CREATE.value,
        Permission.VERSION_VIEW.value,
        Permission.VERSION_CREATE.value,
    },
    
    # 5. MANAGER
    "MANAGER": {
        Permission.PROJECT_VIEW.value,
        Permission.DOCUMENT_VIEW.value,
        Permission.DOCUMENT_UPLOAD.value,
        Permission.DISCOVERY_CHAT.value,
        Permission.REQUIREMENTS_VIEW.value,
        Permission.STAKEHOLDERS_VIEW.value,
        Permission.ANALYSIS_VIEW.value,
        Permission.GAP_ANALYSIS_VIEW.value,
        Permission.RECOMMENDATION_VIEW.value,
        Permission.RECOMMENDATION_APPROVE.value,
        Permission.PROCESS_VIEW.value,
        Permission.ARCHITECTURE_VIEW.value,
        Permission.DATABASE_VIEW.value,
        Permission.API_VIEW.value,
        Permission.UX_VIEW.value,
        Permission.PLANNING_VIEW.value,
        Permission.RISK_VIEW.value,
        Permission.SCORE_VIEW.value,
        Permission.SIMULATION_RUN.value,
        Permission.BLUEPRINT_VIEW.value,
        Permission.BLUEPRINT_APPROVE.value,
        Permission.EXPORT_CREATE.value,
        Permission.COMMENT_CREATE.value,
        Permission.VERSION_VIEW.value,
        Permission.AUDIT_VIEW.value,
    },
    
    # 6. MEMBER
    "MEMBER": {
        Permission.PROJECT_VIEW.value,
        Permission.DOCUMENT_VIEW.value,
        Permission.DOCUMENT_UPLOAD.value,
        Permission.DISCOVERY_CHAT.value,
        Permission.REQUIREMENTS_VIEW.value,
        Permission.STAKEHOLDERS_VIEW.value,
        Permission.ANALYSIS_VIEW.value,
        Permission.GAP_ANALYSIS_VIEW.value,
        Permission.RECOMMENDATION_VIEW.value,
        Permission.PROCESS_VIEW.value,
        Permission.ARCHITECTURE_VIEW.value,
        Permission.DATABASE_VIEW.value,
        Permission.API_VIEW.value,
        Permission.UX_VIEW.value,
        Permission.PLANNING_VIEW.value,
        Permission.RISK_VIEW.value,
        Permission.SCORE_VIEW.value,
        Permission.SIMULATION_RUN.value,
        Permission.BLUEPRINT_VIEW.value,
        Permission.EXPORT_CREATE.value,
        Permission.COMMENT_CREATE.value,
        Permission.VERSION_VIEW.value,
    },
    
    # 7. VIEWER (Read-Only)
    "VIEWER": {
        Permission.PROJECT_VIEW.value,
        Permission.DOCUMENT_VIEW.value,
        Permission.DISCOVERY_CHAT.value,
        Permission.REQUIREMENTS_VIEW.value,
        Permission.STAKEHOLDERS_VIEW.value,
        Permission.ANALYSIS_VIEW.value,
        Permission.GAP_ANALYSIS_VIEW.value,
        Permission.RECOMMENDATION_VIEW.value,
        Permission.PROCESS_VIEW.value,
        Permission.ARCHITECTURE_VIEW.value,
        Permission.DATABASE_VIEW.value,
        Permission.API_VIEW.value,
        Permission.UX_VIEW.value,
        Permission.PLANNING_VIEW.value,
        Permission.RISK_VIEW.value,
        Permission.SCORE_VIEW.value,
        Permission.BLUEPRINT_VIEW.value,
        Permission.EXPORT_CREATE.value,
        Permission.VERSION_VIEW.value,
    }
}

# Add Aliases
ROLE_PERMISSIONS["OWNER"] = ROLE_PERMISSIONS["PROJECT_OWNER"]
ROLE_PERMISSIONS["ANALYST"] = ROLE_PERMISSIONS["BUSINESS_ANALYST"]
ROLE_PERMISSIONS["ARCHITECT"] = ROLE_PERMISSIONS["SOLUTION_ARCHITECT"]

def has_permission(role: str, permission: Union[Permission, str]) -> bool:
    """Checks if the given role has the requested permission."""
    role_key = role.upper() if role else "VIEWER"
    permissions = ROLE_PERMISSIONS.get(role_key, set())
    perm_val = permission.value if isinstance(permission, Permission) else str(permission)
    return perm_val in permissions
